#!/opt/rocks/bin/perl

use strict;
use warnings;

use Getopt::Long;
use lib '/home/whitty/SVN/lib';
use NCBITaxonomyTree;

my $tree = new NCBITaxonomyTree('path' => '/home/whitty/db/NCBI');

my ($method, $species, $input, $output);

GetOptions(
    'input|i=s'     =>  \$input,
    'output|o=s'    =>  \$output,
    'method|m=s'    =>  \$method,
    'species|s=s'   =>  \$species,
);

my $infh;
my $outfh;

my $species_name;
my $taxon_id;

if ($species) {
    $species_name = $species;
    $taxon_id = $tree->get_taxon_id($species_name);
}

if (defined($input)) {
    open $infh, '<', $input || die "Failed to open '$input' for reading: $!";
} else {
    $infh = \*STDIN;
}
if (defined($output)) {
    open $outfh, '>', $output || die "Failed to open '$output' for writing: $!";
} else {
    $outfh = \*STDOUT;
}

while (<$infh>) {
    chomp;

    ## auto-defined method if search was run on PUTs
    if (/^# Generated by GMAP.*\/([^\/]+\S+\.fasta)$/) {
        my $infile = $1;
        if ($infile =~ /\.PUT\./) {
            $infile =~ /^([^\.]+)/;
            my $species_name = $1;
            $species_name =~ s/subsp_/subsp\./;
            $species_name =~ s/var_/var\./;
            $species_name =~ s/_/ /g;

            $taxon_id = $tree->get_taxon_id($species_name);
        }
        if (! defined($method)) {
            $method = 'gmap_'.$taxon_id;
        }
    }

    my @t = split("\t");

    if (scalar(@t) != 9) {
        print $outfh $_."\n";
        next;
    }

    $t[8] =~ s/Identity=/identity=/;
    $t[8] =~ s/Coverage=/coverage=/;

    if (defined($method)) {
        $t[1] = $method;
    }
    #if ($t[2] =~ /^gene$|^mRNA$/ && $t[8] =~ /ID=PUT-\d+[a-z]+-([^-]+)/) {
    if ($t[2] =~ /^gene$/) {
        if ($t[8] =~ /ID=PUT-\d+[a-z]+-([^-]+)/) {
            $species_name = $1;
            $species_name =~ s/subsp_/subsp\./;
            $species_name =~ s/_/ /g;

            $taxon_id = $tree->get_taxon_id($species_name);
        }
        if (defined($species_name) && defined($taxon_id)) {
            $t[8] =~ s/;$//;
            $t[8] .= ';Dbxref=taxon:'.$taxon_id.';organism='.$species_name;
        }
    }
    print $outfh join("\t", @t)."\n";
}
